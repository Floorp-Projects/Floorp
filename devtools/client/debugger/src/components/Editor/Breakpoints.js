/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import Breakpoint from "./Breakpoint";

import { getSelectedSource, getFirstVisibleBreakpoints } from "../../selectors";
import { makeBreakpointId } from "../../utils/breakpoint";
import { connect } from "../../utils/connect";
import { breakpointItemActions } from "./menus/breakpoints";
import { editorItemActions } from "./menus/editor";

class Breakpoints extends Component {
  render() {
    const {
      cx,
      breakpoints,
      selectedSource,
      editor,
      breakpointActions,
      editorActions,
    } = this.props;

    if (!selectedSource || !breakpoints || selectedSource.isBlackBoxed) {
      return null;
    }

    return (
      <div>
        {breakpoints.map(bp => {
          return (
            <Breakpoint
              cx={cx}
              key={makeBreakpointId(bp.location)}
              breakpoint={bp}
              selectedSource={selectedSource}
              editor={editor}
              breakpointActions={breakpointActions}
              editorActions={editorActions}
            />
          );
        })}
      </div>
    );
  }
}

export default connect(
  state => ({
    // Retrieves only the first breakpoint per line so that the
    // breakpoint marker represents only the first breakpoint
    breakpoints: getFirstVisibleBreakpoints(state),
    selectedSource: getSelectedSource(state),
  }),
  dispatch => ({
    breakpointActions: breakpointItemActions(dispatch),
    editorActions: editorItemActions(dispatch),
  })
)(Breakpoints);
