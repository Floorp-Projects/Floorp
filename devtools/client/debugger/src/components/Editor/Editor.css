/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.editor-wrapper {
  --debug-line-border: rgb(145, 188, 219);
  --debug-expression-background: rgba(202, 227, 255, 0.5);
  --debug-line-error-border: rgb(255, 0, 0);
  --debug-expression-error-background: rgba(231, 116, 113, 0.3);
  --line-exception-background: hsl(344, 73%, 97%);
  --highlight-line-duration: 5000ms;
}

.theme-dark .editor-wrapper {
  --debug-expression-background: rgba(202, 227, 255, 0.3);
  --debug-line-border: #7786a2;
  --line-exception-background: hsl(345, 23%, 24%);
}

.editor-wrapper .CodeMirror-linewidget {
  margin-right: -7px;
}

.editor-wrapper {
  min-width: 0 !important;
}

.CodeMirror.cm-s-mozilla,
.CodeMirror-scroll,
.CodeMirror-sizer {
  overflow-anchor: none;
}

/* Prevents inline preview from shifting source height (#1576163) */
.CodeMirror-linewidget {
  padding: 0;
  display: flow-root;
}

/**
 * There's a known codemirror flex issue with chrome that this addresses.
 * BUG https://github.com/firefox-devtools/debugger/issues/63
 */
.editor-wrapper {
  position: absolute;
  width: calc(100% - 1px);
  top: var(--editor-header-height);
  bottom: var(--editor-footer-height);
  left: 0px;
}

html[dir="rtl"] .editor-mount {
  direction: ltr;
}

.theme-light .cm-s-mozilla .empty-line .CodeMirror-linenumber {
  color: var(--grey-40);
}

.theme-dark .cm-s-mozilla .empty-line .CodeMirror-linenumber {
  color: var(--grey-50);
}

.function-search {
  max-height: 300px;
  overflow: hidden;
}

.function-search .results {
  height: auto;
}

.editor.hit-marker {
  height: 15px;
}

.editor-wrapper .highlight-lines {
  background: var(--theme-selection-background-hover);
}

.CodeMirror {
  width: 100%;
  height: 100%;
}

.editor-wrapper .editor-mount {
  width: 100%;
  background-color: var(--theme-body-background);
  font-size: var(--theme-code-font-size);
  line-height: var(--theme-code-line-height);
}

/* set the linenumber white when there is a breakpoint */
.editor-wrapper:not(.skip-pausing)
  .new-breakpoint
  .CodeMirror-gutter-wrapper
  .CodeMirror-linenumber {
  color: white;
}

/* move the breakpoint below the other gutter elements */
.new-breakpoint .CodeMirror-gutter-elt:nth-child(2) {
  z-index: 0;
}

.theme-dark .editor-wrapper .CodeMirror-line .cm-comment {
  color: var(--theme-comment);
}

.debug-expression {
  background-color: var(--debug-expression-background);
  border-style: solid;
  border-color: var(--debug-expression-background);
  border-width: 1px 0px 1px 0px;
  position: relative;
}

.debug-expression::before {
  content: "";
  line-height: 1px;
  border-top: 1px solid var(--blue-50);
  background: transparent;
  position: absolute;
  top: -2px;
  left: 0px;
  width: 100%;
 }

.debug-expression::after {
  content: "";
  line-height: 1px;
  border-bottom: 1px solid var(--blue-50);
  position: absolute;
  bottom: -2px;
  left: 0px;
  width: 100%;
 }

.to-line-end ~ .CodeMirror-widget {
  background-color: var(--debug-expression-background);
}

.debug-expression-error {
  background-color: var(--debug-expression-error-background);
}

.new-debug-line > .CodeMirror-line {
  background-color: transparent !important;
  outline: var(--debug-line-border) solid 1px;
}

/* Don't display the highlight color since the debug line
   is already highlighted */
.new-debug-line .CodeMirror-activeline-background {
  display: none;
}

.new-debug-line-error > .CodeMirror-line {
  background-color: var(--debug-expression-error-background) !important;
  outline: var(--debug-line-error-border) solid 1px;
}

/* Don't display the highlight color since the debug line
  is already highlighted */
.new-debug-line-error .CodeMirror-activeline-background {
  display: none;
}
.highlight-line .CodeMirror-line {
  animation-name: fade-highlight-out;
  animation-duration: var(--highlight-line-duration);
  animation-timing-function: ease-out;
  animation-direction: forwards;
}

@keyframes fade-highlight-out {
  0% {
    background-color: var(--theme-contrast-background);
  }
  30% {
    background-color: var(--theme-contrast-background);
  }
  100% {
    background-color: transparent;
  }
}

.visible {
  visibility: visible;
}

/* Code folding */
.editor-wrapper .CodeMirror-foldgutter-open {
  color: var(--grey-40);
}

.editor-wrapper .CodeMirror-foldgutter-open,
.editor-wrapper .CodeMirror-foldgutter-folded {
  fill: var(--grey-40);
}

.editor-wrapper .CodeMirror-foldgutter-open::before,
.editor-wrapper .CodeMirror-foldgutter-open::after {
  border-top: none;
}

.editor-wrapper .CodeMirror-foldgutter-folded::before,
.editor-wrapper .CodeMirror-foldgutter-folded::after {
  border-left: none;
}

.editor-wrapper .CodeMirror-foldgutter .CodeMirror-guttermarker-subtle {
  visibility: visible;
}

.editor-wrapper .CodeMirror-foldgutter .CodeMirror-linenumber {
  text-align: left;
  padding: 0 0 0 2px;
}

/* Exception line */
.line-exception {
  background-color: var(--line-exception-background);
}

.mark-text-exception {
  text-decoration: var(--red-50) wavy underline;
  text-decoration-skip-ink: none;
}
