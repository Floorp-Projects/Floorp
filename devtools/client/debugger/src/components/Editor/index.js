/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import PropTypes from "prop-types";
import React, { PureComponent } from "react";
import { bindActionCreators } from "redux";
import ReactDOM from "react-dom";
import { connect } from "../../utils/connect";
import classnames from "classnames";

import { getLineText } from "./../../utils/source";
import { features } from "../../utils/prefs";
import { getIndentation } from "../../utils/indentation";

import { showMenu } from "../../context-menu/menu";
import {
  createBreakpointItems,
  breakpointItemActions,
} from "./menus/breakpoints";

import { continueToHereItem, editorItemActions } from "./menus/editor";

import {
  getActiveSearch,
  getSelectedLocation,
  getSelectedSourceWithContent,
  getConditionalPanelLocation,
  getSymbols,
  getIsCurrentThreadPaused,
  getCurrentThread,
  getThreadContext,
  getSkipPausing,
  getInlinePreview,
  getEditorWrapping,
  getHighlightedCalls,
} from "../../selectors";

// Redux actions
import actions from "../../actions";

import SearchBar from "./SearchBar";
import HighlightLines from "./HighlightLines";
import Preview from "./Preview";
import Breakpoints from "./Breakpoints";
import ColumnBreakpoints from "./ColumnBreakpoints";
import DebugLine from "./DebugLine";
import HighlightLine from "./HighlightLine";
import EmptyLines from "./EmptyLines";
import EditorMenu from "./EditorMenu";
import ConditionalPanel from "./ConditionalPanel";
import InlinePreviews from "./InlinePreviews";
import HighlightCalls from "./HighlightCalls";
import Exceptions from "./Exceptions";

import {
  showSourceText,
  showLoading,
  showErrorMessage,
  getEditor,
  clearEditor,
  getCursorLine,
  getCursorColumn,
  lineAtHeight,
  toSourceLine,
  getDocument,
  scrollToColumn,
  toEditorPosition,
  getSourceLocationFromMouseEvent,
  hasDocument,
  onMouseOver,
  startOperation,
  endOperation,
} from "../../utils/editor";

import { resizeToggleButton, resizeBreakpointGutter } from "../../utils/ui";
import Services from "devtools-services";

const { debounce } = require("devtools/shared/debounce");

const { appinfo } = Services;
const isMacOS = appinfo.OS === "Darwin";

function isSecondary(ev) {
  return isMacOS && ev.ctrlKey && ev.button === 0;
}

function isCmd(ev) {
  return isMacOS ? ev.metaKey : ev.ctrlKey;
}

import "./Editor.css";
import "./Breakpoints.css";
import "./InlinePreview.css";

const cssVars = {
  searchbarHeight: "var(--editor-searchbar-height)",
};

class Editor extends PureComponent {
  $editorWrapper;
  constructor(props) {
    super(props);

    this.state = {
      highlightedLineRange: null,
      editor: null,
      contextMenu: null,
    };
  }

  componentWillReceiveProps(nextProps) {
    let { editor } = this.state;

    if (!editor && nextProps.selectedSource) {
      editor = this.setupEditor();
    }

    startOperation();
    this.setText(nextProps, editor);
    this.setSize(nextProps, editor);
    this.scrollToLocation(nextProps, editor);
    endOperation();

    if (this.props.selectedSource != nextProps.selectedSource) {
      this.props.updateViewport();
      resizeBreakpointGutter(editor.codeMirror);
      resizeToggleButton(editor.codeMirror);
    }
  }

  setupEditor() {
    const editor = getEditor();

    // disables the default search shortcuts
    editor._initShortcuts = () => {};

    const node = ReactDOM.findDOMNode(this);
    if (node instanceof HTMLElement) {
      editor.appendToLocalElement(node.querySelector(".editor-mount"));
    }

    const { codeMirror } = editor;
    const codeMirrorWrapper = codeMirror.getWrapperElement();

    codeMirror.on("gutterClick", this.onGutterClick);

    if (features.commandClick) {
      document.addEventListener("keydown", this.commandKeyDown);
      document.addEventListener("keyup", this.commandKeyUp);
    }

    // Set code editor wrapper to be focusable
    codeMirrorWrapper.tabIndex = 0;
    codeMirrorWrapper.addEventListener("keydown", e => this.onKeyDown(e));
    codeMirrorWrapper.addEventListener("click", e => this.onClick(e));
    codeMirrorWrapper.addEventListener("mouseover", onMouseOver(codeMirror));

    const toggleFoldMarkerVisibility = e => {
      if (node instanceof HTMLElement) {
        node
          .querySelectorAll(".CodeMirror-guttermarker-subtle")
          .forEach(elem => {
            elem.classList.toggle("visible");
          });
      }
    };

    const codeMirrorGutter = codeMirror.getGutterElement();
    codeMirrorGutter.addEventListener("mouseleave", toggleFoldMarkerVisibility);
    codeMirrorGutter.addEventListener("mouseenter", toggleFoldMarkerVisibility);
    codeMirrorWrapper.addEventListener("contextmenu", event =>
      this.openMenu(event)
    );

    codeMirror.on("scroll", this.onEditorScroll);
    this.onEditorScroll();
    this.setState({ editor });
    return editor;
  }

  componentDidMount() {
    const { shortcuts } = this.context;

    shortcuts.on(L10N.getStr("toggleBreakpoint.key"), this.onToggleBreakpoint);
    shortcuts.on(
      L10N.getStr("toggleCondPanel.breakpoint.key"),
      this.onToggleConditionalPanel
    );
    shortcuts.on(
      L10N.getStr("toggleCondPanel.logPoint.key"),
      this.onToggleConditionalPanel
    );
    shortcuts.on(
      L10N.getStr("sourceTabs.closeTab.key"),
      this.onCloseShortcutPress
    );
    shortcuts.on("Esc", this.onEscape);
  }

  onCloseShortcutPress = e => {
    const { cx, selectedSource } = this.props;
    if (selectedSource) {
      e.preventDefault();
      e.stopPropagation();
      this.props.closeTab(cx, selectedSource, "shortcut");
    }
  };

  componentWillUnmount() {
    const { editor } = this.state;
    if (editor) {
      editor.destroy();
      editor.codeMirror.off("scroll", this.onEditorScroll);
      this.setState({ editor: null });
    }

    const { shortcuts } = this.context;
    shortcuts.off(L10N.getStr("sourceTabs.closeTab.key"));
    shortcuts.off(L10N.getStr("toggleBreakpoint.key"));
    shortcuts.off(L10N.getStr("toggleCondPanel.breakpoint.key"));
    shortcuts.off(L10N.getStr("toggleCondPanel.logPoint.key"));
  }

  getCurrentLine() {
    const { codeMirror } = this.state.editor;
    const { selectedSource } = this.props;
    if (!selectedSource) {
      return;
    }

    const line = getCursorLine(codeMirror);
    return toSourceLine(selectedSource.id, line);
  }

  onToggleBreakpoint = e => {
    e.preventDefault();
    e.stopPropagation();

    const line = this.getCurrentLine();
    if (typeof line !== "number") {
      return;
    }

    this.props.toggleBreakpointAtLine(this.props.cx, line);
  };

  onToggleConditionalPanel = e => {
    e.stopPropagation();
    e.preventDefault();

    const {
      conditionalPanelLocation,
      closeConditionalPanel,
      openConditionalPanel,
      selectedSource,
    } = this.props;

    const line = this.getCurrentLine();

    const { codeMirror } = this.state.editor;
    // add one to column for correct position in editor.
    const column = getCursorColumn(codeMirror) + 1;

    if (conditionalPanelLocation) {
      return closeConditionalPanel();
    }

    if (!selectedSource) {
      return;
    }

    if (typeof line !== "number") {
      return;
    }

    return openConditionalPanel(
      {
        line,
        column,
        sourceId: selectedSource.id,
      },
      false
    );
  };

  onEditorScroll = debounce(this.props.updateViewport, 75);

  commandKeyDown = e => {
    const { key } = e;
    if (this.props.isPaused && key === "Meta") {
      const { cx, highlightCalls } = this.props;
      highlightCalls(cx);
    }
  };

  commandKeyUp = e => {
    const { key } = e;
    if (key === "Meta") {
      const { cx, unhighlightCalls } = this.props;
      unhighlightCalls(cx);
    }
  };

  onKeyDown(e) {
    const { codeMirror } = this.state.editor;
    const { key, target } = e;
    const codeWrapper = codeMirror.getWrapperElement();
    const textArea = codeWrapper.querySelector("textArea");

    if (key === "Escape" && target == textArea) {
      e.stopPropagation();
      e.preventDefault();
      codeWrapper.focus();
    } else if (key === "Enter" && target == codeWrapper) {
      e.preventDefault();
      // Focus into editor's text area
      textArea.focus();
    }
  }

  /*
   * The default Esc command is overridden in the CodeMirror keymap to allow
   * the Esc keypress event to be catched by the toolbox and trigger the
   * split console. Restore it here, but preventDefault if and only if there
   * is a multiselection.
   */
  onEscape = e => {
    if (!this.state.editor) {
      return;
    }

    const { codeMirror } = this.state.editor;
    if (codeMirror.listSelections().length > 1) {
      codeMirror.execCommand("singleSelection");
      e.preventDefault();
    }
  };

  openMenu(event) {
    event.stopPropagation();
    event.preventDefault();

    const {
      cx,
      selectedSource,
      breakpointActions,
      editorActions,
      isPaused,
      conditionalPanelLocation,
      closeConditionalPanel,
    } = this.props;
    const { editor } = this.state;
    if (!selectedSource || !editor) {
      return;
    }

    // only allow one conditionalPanel location.
    if (conditionalPanelLocation) {
      closeConditionalPanel();
    }

    const target = event.target;
    const { id: sourceId } = selectedSource;
    const line = lineAtHeight(editor, sourceId, event);

    if (typeof line != "number") {
      return;
    }

    const location = { line, column: undefined, sourceId };

    if (target.classList.contains("CodeMirror-linenumber")) {
      const lineText = getLineText(
        sourceId,
        selectedSource.content,
        line
      ).trim();

      return showMenu(event, [
        ...createBreakpointItems(cx, location, breakpointActions, lineText),
        { type: "separator" },
        continueToHereItem(cx, location, isPaused, editorActions),
      ]);
    }

    if (target.getAttribute("id") === "columnmarker") {
      return;
    }

    this.setState({ contextMenu: event });
  }

  clearContextMenu = () => {
    this.setState({ contextMenu: null });
  };

  onGutterClick = (cm, line, gutter, ev) => {
    const {
      cx,
      selectedSource,
      conditionalPanelLocation,
      closeConditionalPanel,
      addBreakpointAtLine,
      continueToHere,
      toggleBlackBox,
    } = this.props;

    // ignore right clicks in the gutter
    if (isSecondary(ev) || ev.button === 2 || !selectedSource) {
      return;
    }

    // if user clicks gutter to set breakpoint on blackboxed source, un-blackbox the source.
    if (selectedSource?.isBlackBoxed) {
      toggleBlackBox(cx, selectedSource);
    }

    if (conditionalPanelLocation) {
      return closeConditionalPanel();
    }

    if (gutter === "CodeMirror-foldgutter") {
      return;
    }

    const sourceLine = toSourceLine(selectedSource.id, line);
    if (typeof sourceLine !== "number") {
      return;
    }

    if (isCmd(ev)) {
      return continueToHere(cx, {
        line: sourceLine,
        column: undefined,
        sourceId: selectedSource.id,
      });
    }

    return addBreakpointAtLine(cx, sourceLine, ev.altKey, ev.shiftKey);
  };

  onGutterContextMenu = event => {
    return this.openMenu(event);
  };

  onClick(e) {
    const {
      cx,
      selectedSource,
      updateCursorPosition,
      jumpToMappedLocation,
    } = this.props;

    if (selectedSource) {
      const sourceLocation = getSourceLocationFromMouseEvent(
        this.state.editor,
        selectedSource,
        e
      );

      if (e.metaKey && e.altKey) {
        jumpToMappedLocation(cx, sourceLocation);
      }

      updateCursorPosition(sourceLocation);
    }
  }

  shouldScrollToLocation(nextProps, editor) {
    const { selectedLocation, selectedSource } = this.props;
    if (
      !editor ||
      !nextProps.selectedSource ||
      !nextProps.selectedLocation ||
      !nextProps.selectedLocation.line ||
      !nextProps.selectedSource.content
    ) {
      return false;
    }

    const isFirstLoad =
      (!selectedSource || !selectedSource.content) &&
      nextProps.selectedSource.content;
    const locationChanged = selectedLocation !== nextProps.selectedLocation;
    const symbolsChanged = nextProps.symbols != this.props.symbols;

    return isFirstLoad || locationChanged || symbolsChanged;
  }

  scrollToLocation(nextProps, editor) {
    const { selectedLocation, selectedSource } = nextProps;

    if (selectedLocation && this.shouldScrollToLocation(nextProps, editor)) {
      let { line, column } = toEditorPosition(selectedLocation);

      if (selectedSource && hasDocument(selectedSource.id)) {
        const doc = getDocument(selectedSource.id);
        const lineText = doc.getLine(line);
        column = Math.max(column, getIndentation(lineText));
      }

      scrollToColumn(editor.codeMirror, line, column);
    }
  }

  setSize(nextProps, editor) {
    if (!editor) {
      return;
    }

    if (
      nextProps.startPanelSize !== this.props.startPanelSize ||
      nextProps.endPanelSize !== this.props.endPanelSize
    ) {
      editor.codeMirror.setSize();
    }
  }

  setText(props, editor) {
    const { selectedSource, symbols } = props;

    if (!editor) {
      return;
    }

    // check if we previously had a selected source
    if (!selectedSource) {
      return this.clearEditor();
    }

    if (!selectedSource.content) {
      return showLoading(editor);
    }

    if (selectedSource.content.state === "rejected") {
      let { value } = selectedSource.content;
      if (typeof value !== "string") {
        value = "Unexpected source error";
      }

      return this.showErrorMessage(value);
    }

    return showSourceText(
      editor,
      selectedSource,
      selectedSource.content.value,
      symbols
    );
  }

  clearEditor() {
    const { editor } = this.state;
    if (!editor) {
      return;
    }

    clearEditor(editor);
  }

  showErrorMessage(msg) {
    const { editor } = this.state;
    if (!editor) {
      return;
    }

    showErrorMessage(editor, msg);
  }

  getInlineEditorStyles() {
    const { searchOn } = this.props;

    if (searchOn) {
      return {
        height: `calc(100% - ${cssVars.searchbarHeight})`,
      };
    }

    return {
      height: "100%",
    };
  }

  renderItems() {
    const {
      cx,
      selectedSource,
      conditionalPanelLocation,
      isPaused,
      inlinePreviewEnabled,
      editorWrappingEnabled,
    } = this.props;
    const { editor, contextMenu } = this.state;

    if (!selectedSource || !editor || !getDocument(selectedSource.id)) {
      return null;
    }

    return (
      <div>
        <HighlightCalls editor={editor} selectedSource={selectedSource} />
        <DebugLine />
        <HighlightLine />
        <EmptyLines editor={editor} />
        <Breakpoints editor={editor} cx={cx} />
        <Preview editor={editor} editorRef={this.$editorWrapper} />
        <HighlightLines editor={editor} />
        <Exceptions />
        {
          <EditorMenu
            editor={editor}
            contextMenu={contextMenu}
            clearContextMenu={this.clearContextMenu}
            selectedSource={selectedSource}
            editorWrappingEnabled={editorWrappingEnabled}
          />
        }
        {conditionalPanelLocation ? <ConditionalPanel editor={editor} /> : null}
        {features.columnBreakpoints ? (
          <ColumnBreakpoints editor={editor} />
        ) : null}
        {isPaused && inlinePreviewEnabled ? (
          <InlinePreviews editor={editor} selectedSource={selectedSource} />
        ) : null}
      </div>
    );
  }

  renderSearchBar() {
    const { editor } = this.state;

    if (!this.props.selectedSource) {
      return null;
    }

    return <SearchBar editor={editor} />;
  }

  render() {
    const { selectedSource, skipPausing } = this.props;
    return (
      <div
        className={classnames("editor-wrapper", {
          blackboxed: selectedSource?.isBlackBoxed,
          "skip-pausing": skipPausing,
        })}
        ref={c => (this.$editorWrapper = c)}
      >
        <div
          className="editor-mount devtools-monospace"
          style={this.getInlineEditorStyles()}
        />
        {this.renderSearchBar()}
        {this.renderItems()}
      </div>
    );
  }
}

Editor.contextTypes = {
  shortcuts: PropTypes.object,
};

const mapStateToProps = state => {
  const selectedSource = getSelectedSourceWithContent(state);

  return {
    cx: getThreadContext(state),
    selectedLocation: getSelectedLocation(state),
    selectedSource,
    searchOn: getActiveSearch(state) === "file",
    conditionalPanelLocation: getConditionalPanelLocation(state),
    symbols: getSymbols(state, selectedSource),
    isPaused: getIsCurrentThreadPaused(state),
    skipPausing: getSkipPausing(state),
    inlinePreviewEnabled: getInlinePreview(state),
    editorWrappingEnabled: getEditorWrapping(state),
    highlightedCalls: getHighlightedCalls(state, getCurrentThread(state)),
  };
};

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(
    {
      openConditionalPanel: actions.openConditionalPanel,
      closeConditionalPanel: actions.closeConditionalPanel,
      continueToHere: actions.continueToHere,
      toggleBreakpointAtLine: actions.toggleBreakpointAtLine,
      addBreakpointAtLine: actions.addBreakpointAtLine,
      jumpToMappedLocation: actions.jumpToMappedLocation,
      traverseResults: actions.traverseResults,
      updateViewport: actions.updateViewport,
      updateCursorPosition: actions.updateCursorPosition,
      closeTab: actions.closeTab,
      toggleBlackBox: actions.toggleBlackBox,
      highlightCalls: actions.highlightCalls,
      unhighlightCalls: actions.unhighlightCalls,
    },
    dispatch
  ),
  breakpointActions: breakpointItemActions(dispatch),
  editorActions: editorItemActions(dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Editor);
