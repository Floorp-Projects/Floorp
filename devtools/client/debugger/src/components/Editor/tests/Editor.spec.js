/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import Editor from "../index";
import { getDocument } from "../../../utils/editor/source-documents";
import * as asyncValue from "../../../utils/async-value";

function generateDefaults(overrides) {
  return {
    toggleBreakpoint: jest.fn(),
    updateViewport: jest.fn(),
    toggleDisabledBreakpoint: jest.fn(),
    ...overrides,
  };
}

function createMockEditor() {
  return {
    codeMirror: {
      doc: {},
      getOption: jest.fn(),
      setOption: jest.fn(),
      scrollTo: jest.fn(),
      charCoords: ({ line, ch }) => ({ top: line, left: ch }),
      getScrollerElement: () => ({ offsetWidth: 0, offsetHeight: 0 }),
      getScrollInfo: () => ({
        top: 0,
        left: 0,
        clientWidth: 0,
        clientHeight: 0,
      }),
      defaultCharWidth: () => 0,
      defaultTextHeight: () => 0,
      display: { gutters: { querySelector: jest.fn() } },
    },
    setText: jest.fn(),
    on: jest.fn(),
    off: jest.fn(),
    createDocument: () => {
      let val;
      return {
        getLine: line => "",
        getValue: () => val,
        setValue: newVal => (val = newVal),
      };
    },
    replaceDocument: jest.fn(),
    setMode: jest.fn(),
  };
}

function createMockSourceWithContent(overrides) {
  const {
    loadedState = "loaded",
    text = "the text",
    contentType = undefined,
    error = undefined,
    ...otherOverrides
  } = overrides;

  const source = {
    id: "foo",
    url: "foo",
    ...otherOverrides,
  };
  let content = null;
  if (loadedState === "loaded") {
    if (typeof text !== "string") {
      throw new Error("Cannot create a non-text source");
    }

    content = error
      ? asyncValue.rejected(error)
      : asyncValue.fulfilled({
          type: "text",
          value: text,
          contentType: contentType || undefined,
        });
  }

  return {
    ...source,
    content,
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  const mockEditor = createMockEditor();

  const component = shallow(<Editor.WrappedComponent {...props} />, {
    context: {
      shortcuts: { on: jest.fn() },
    },
    disableLifecycleMethods: true,
  });

  return { component, props, mockEditor };
}

describe("Editor", () => {
  describe("When empty", () => {
    it("should render", async () => {
      const { component } = render();
      expect(component).toMatchSnapshot();
    });
  });

  describe("When loading initial source", () => {
    it("should show a loading message", async () => {
      const { component, mockEditor } = render();
      await component.setState({ editor: mockEditor });
      component.setProps({
        selectedSource: {
          source: { loadedState: "loading" },
          content: null,
        },
      });

      expect(mockEditor.replaceDocument.mock.calls[0][0].getValue()).toBe(
        "Loading…"
      );
      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([]);
    });
  });

  describe("When loaded", () => {
    it("should show text", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          loadedState: "loaded",
        }),
        selectedLocation: { sourceId: "foo", line: 3, column: 1 },
      });

      expect(mockEditor.setText.mock.calls).toEqual([["the text"]]);
      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([[1, 2]]);
    });
  });

  describe("When error", () => {
    it("should show error text", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          loadedState: "loaded",
          text: undefined,
          error: "error text",
        }),
        selectedLocation: { sourceId: "bad-foo", line: 3, column: 1 },
      });

      expect(mockEditor.setText.mock.calls).toEqual([
        ["Error loading this URI: error text"],
      ]);
    });

    it("should show wasm error", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          loadedState: "loaded",
          isWasm: true,
          text: undefined,
          error: "blah WebAssembly binary source is not available blah",
        }),
        selectedLocation: { sourceId: "bad-foo", line: 3, column: 1 },
      });

      expect(mockEditor.setText.mock.calls).toEqual([
        ["Please refresh to debug this module"],
      ]);
    });
  });

  describe("When navigating to a loading source", () => {
    it("should show loading message and not scroll", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          loadedState: "loaded",
        }),
        selectedLocation: { sourceId: "foo", line: 3, column: 1 },
      });

      // navigate to a new source that is still loading
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          id: "bar",
          loadedState: "loading",
        }),
        selectedLocation: { sourceId: "bar", line: 1, column: 1 },
      });

      expect(mockEditor.replaceDocument.mock.calls[1][0].getValue()).toBe(
        "Loading…"
      );

      expect(mockEditor.setText.mock.calls).toEqual([["the text"]]);

      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([[1, 2]]);
    });

    it("should set the mode when symbols load", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });

      const selectedSource = createMockSourceWithContent({
        loadedState: "loaded",
        contentType: "javascript",
      });

      await component.setProps({ ...props, selectedSource });

      const symbols = { hasJsx: true };
      await component.setProps({
        ...props,
        selectedSource,
        symbols,
      });

      expect(mockEditor.setMode.mock.calls).toEqual([
        [{ name: "javascript" }],
        [{ name: "jsx" }],
      ]);
    });

    it("should not re-set the mode when the location changes", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });

      const selectedSource = createMockSourceWithContent({
        loadedState: "loaded",
        contentType: "javascript",
      });

      await component.setProps({ ...props, selectedSource });

      // symbols are parsed
      const symbols = { hasJsx: true };
      await component.setProps({
        ...props,
        selectedSource,
        symbols,
      });

      // selectedLocation changes e.g. pausing/stepping
      mockEditor.codeMirror.doc = getDocument(selectedSource.id);
      mockEditor.codeMirror.getOption = () => ({ name: "jsx" });
      const selectedLocation = { sourceId: "foo", line: 4, column: 1 };

      await component.setProps({
        ...props,
        selectedSource,
        symbols,
        selectedLocation,
      });

      expect(mockEditor.setMode.mock.calls).toEqual([
        [{ name: "javascript" }],
        [{ name: "jsx" }],
      ]);
    });
  });

  describe("When navigating to a loaded source", () => {
    it("should show text and then scroll", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          loadedState: "loading",
        }),
        selectedLocation: { sourceId: "foo", line: 1, column: 1 },
      });

      // navigate to a new source that is still loading
      await component.setProps({
        ...props,
        selectedSource: createMockSourceWithContent({
          loadedState: "loaded",
        }),
        selectedLocation: { sourceId: "foo", line: 1, column: 1 },
      });

      expect(mockEditor.replaceDocument.mock.calls[0][0].getValue()).toBe(
        "Loading…"
      );

      expect(mockEditor.setText.mock.calls).toEqual([["the text"]]);

      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([[1, 0]]);
    });
  });
});
