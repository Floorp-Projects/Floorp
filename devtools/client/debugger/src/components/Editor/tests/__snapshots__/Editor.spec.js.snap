// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor When empty should render 1`] = `
<div
  className="editor-wrapper"
>
  <div
    className="editor-mount devtools-monospace"
    style={
      Object {
        "height": "100%",
      }
    }
  />
</div>
`;
